fdbserver/GlobalTagThrottler.actor.cpp

```diff
@@ -27,197 +27,495 @@
 					}
-
+					self->removeUnseenQuotas(tagsWithQuota);
 					++self->throttledTagChangeId;
-					// FIXME: Should wait on watch instead
-					// wait(tr.watch(tagThrottleSignalKey));
 					wait(delay(5.0));
 					TraceEvent("GlobalTagThrottler_ChangeSignaled");
 					CODE_PROBE(true, "Global tag throttler detected quota changes");
 					break;
 				} catch (Error& e) {
-					TraceEvent("GlobalTagThrottlerMonitoringChangesError", self->id).error(e);
+					TraceEvent("GlobalTagThrottler_MonitoringChangesError", self->id).error(e);
 					wait(tr.onError(e));
 				}
 			}
 		}
 	}
 
 public:
-	GlobalTagThrottlerImpl(Database db, UID id) : db(db), id(id) { traceActor = tracer(this); }
+	GlobalTagThrottlerImpl(Database db, UID id) : db(db), id(id) {}
 	Future<Void> monitorThrottlingChanges() { return monitorThrottlingChanges(this); }
-	void addRequests(TransactionTag tag, int count) { trackedTags[tag].addTransactions(count); }
+	void addRequests(TransactionTag tag, int count) { tagStatistics[tag].addTransactions(count); }
 	uint64_t getThrottledTagChangeId() const { return throttledTagChangeId; }
 	PrioritizedTransactionTagMap<ClientTagThrottleLimits> getClientRates() {
-		// TODO: For now, only enforce total throttling rates.
-		// We should use reserved quotas as well.
 		PrioritizedTransactionTagMap<ClientTagThrottleLimits> result;
-		for (auto& [tag, quotaAndCounters] : trackedTags) {
+		lastBusyReadTagCount = lastBusyWriteTagCount = 0;
+		for (auto& [tag, stats] : tagStatistics) {
 			// Currently there is no differentiation between batch priority and default priority transactions
-			auto const limit = quotaAndCounters.updateAndGetPerClientLimit();
-			if (limit.present()) {
-				result[TransactionPriority::BATCH][tag] = result[TransactionPriority::DEFAULT][tag] = limit.get();
+			auto const readLimitingTps = getLimitingTps(tag, OpType::READ);
+			auto const writeLimitingTps = getLimitingTps(tag, OpType::WRITE);
+			Optional<double> limitingTps;
+			if (readLimitingTps.present() && writeLimitingTps.present()) {
+				limitingTps = std::min(readLimitingTps.get(), writeLimitingTps.get());
+			} else if (readLimitingTps.present()) {
+				limitingTps = readLimitingTps;
+			} else {
+				limitingTps = writeLimitingTps;
+			}
+
+			auto const averageTransactionReadCost = getAverageTransactionCost(tag, OpType::READ);
+			auto const averageTransactionWriteCost = getAverageTransactionCost(tag, OpType::WRITE);
+			auto const readDesiredTps = getDesiredTps(tag, OpType::READ, averageTransactionReadCost);
+			auto const writeDesiredTps = getDesiredTps(tag, OpType::WRITE, averageTransactionWriteCost);
+			Optional<double> desiredTps;
+			if (readDesiredTps.present() && writeDesiredTps.present()) {
+				desiredTps = std::min(readDesiredTps.get(), writeDesiredTps.get());
+			} else if (readDesiredTps.present()) {
+				desiredTps = readDesiredTps;
+			} else {
+				desiredTps = writeDesiredTps;
+			}
+
+			if (!desiredTps.present()) {
+				continue;
+			}
```

Comment:
Does this mean if there's no meaningful `desiredTps` returned, the `limitingTps` also doesn't matter? 
I think we still need to set the client rate to limitingTps here.

Reply:
If there is no `desiredTps`, then there is no quota set, so the global tag throttling is inactive for this tag. There is no way to compute `limitingTps` without a quota set.